wagtail/snippets/views/snippets.py

```diff
@@ -191,17 +191,16 @@ def get_context_data(self, **kwargs):
 
     def get_template_names(self):
         if self.results_only:
-            return ["wagtailsnippets/snippets/index_results.html"]
+            return self.viewset.get_index_results_template()
```

Comment:
The ship sails from this PR onwards, so as long as this hasn't been merged, there's still time to change the approach 😄

I strongly agree with your points there, and in fact I considered extracting the `get_templates()` into a mixin as it makes a lot of sense for this use case. However, this approach doesn't allow you to do the customisation through `SnippetViewSet`, so I decided to copy how ModelAdmin does it.

How do you think such method overrides could be supported? Or do we just not support it, and say that "if you need to do such customisations then you'll need to make subclasses of the view classes with your own `get_templates()`, and override all the `foo_view_class` attribute on the viewset to use those subclasses"?

Or, do we possibly add something like a `SnippetViewSet.get_templates_mixin` and apply the mixin dynamically with `type()`, e.g.

```py
@property
def index_view_class(self):
    return type('IndexView', (self.get_templates_mixin, IndexView))
```

So developers can define their own mixin with their own `get_templates()` logic, then only swap `SnippetViewSet.get_templates_mixin`?

That does mean we'll need to change all the `foo_view_class = ...` for the affected views into a `@property`, though (no big deal).

Also, does this mean we don't want to implement ModelAdmin's view-specific methods like `ModelAdmin.get_index_results_template()` that ModelAdmin supports?

Reply:
OK, I managed to avoid this pattern by allowing the view's `template_name` to be a list. Don't know why Django's `TemplateResponseMixin` always forcefully wrap it in a list without checking. Even if `get_template_names()` returns a string (instead of a list as the docstring says it must), it still works fine as long as the `response_class` resolves it correctly (which `SimpleTemplateResponse` does by default).

Anyway, the new approach is much cleaner and lets all the logic be encapsulated within `SnippetViewSet`, and all the views know is that the `template_name` has been overridden. Using the views without the viewset should still work because they still have the default `template_name`s, and they're only overridden through `SnippetViewSet` (or if the view is subclassed).

That said, removing the viewset from the views makes it impossible to override things like the `get_queryset()` methods on a per-request basis (like ModelAdmin supports), unless we want to pass the methods themselves to the views... 🤔